### PR TITLE
Vue 3

### PR DESCRIPTION
**Pull request by martinmaillard**

### CODE REVIEW

File: ui/src/forms/editors/DictionaryTableEditor.vue [hunk 1]
```diff
@@ -39,18 +36,18 @@
...
-import { defineComponent, PropType } from '@vue/composition-api'
+import { defineComponent, PropType } from 'vue'
 import { PropertyObjectState } from '@hydrofoil/shaperone-core/models/forms'
 import { prov, sh } from '@tpluscode/rdf-ns-builders'
 import { Term } from 'rdf-js'
 import { PropertyShape } from '@rdfine/shacl'
 import { PropertyRenderer } from '@hydrofoil/shaperone-core/renderer'
 import RadioButton from '@/components/RadioButton.vue'
-import RenderWcTemplate from '../RenderWcTemplate.vue'
+import DictionaryTableEditorPair from './DictionaryTableEditorPair.vue'
```
Comment (tpluscode) [inline ui/src/forms/editors/DictionaryTableEditor.vue:46]: Why did you add an additional component if it's used only once here?

File: ui/src/forms/vue-custom-element.ts [hunk 2]
```diff
@@ -0,0 +1,481 @@
+// Copied from this PR https://github.com/vuejs/core/pull/4404
+// that adds a `{ shadowRoot: false }` option to `defineCustomElement`.
+// I also force-set the `appContext` on created components so they can
+// access registered and configured Oruga components.
```
Comment (tpluscode) [inline ui/src/forms/vue-custom-element.ts:4]: We need our own copy because of that shadow dom option?

**Review by tpluscode — COMMENTED** [review]
I think there isn't much one can tell by just looking at code. I left some comments but since the app appears to function correctly I think it's good

That said, I noticed one error in the console which appear exactly when you open a hierarchy for editing and then switch to `Cube Creator` or `Shared dimensions`

```
index.ts?0260:59 Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'split')
    at _callee$ (index.ts?0260:59:1)
    at tryCatch (runtime.js?74a1:63:1)
    at Generator.invoke [as _invoke] (runtime.js?74a1:294:1)
    at Generator.eval [as next] (runtime.js?74a1:119:1)
    at asyncGeneratorStep (asyncToGenerator.js?b7e0:3:1)
    at _next (asyncToGenerator.js?b7e0:25:1)
    at eval (asyncToGenerator.js?b7e0:32:1)
    at new Promise (<anonymous>)
    at eval (asyncToGenerator.js?b7e0:21:1)
    at Object.fetchResource (index.ts?0260:56:1)
```

File: ui/src/components/JobForm.vue [hunk 3]
```diff
@@ -52,12 +52,17 @@ export default defineComponent({
     },
   },
 
-  data (): { resource: GraphPointer | null, shape: Shape | null, error: ErrorDetails | null, isSubmitting: boolean } {
+  setup () {
+    const resource: Ref<GraphPointer | null> = ref(null)
+    const shape: Ref<Shape | null> = ref(null)
+    const error: Ref<ErrorDetails | null> = ref(null)
+    const isSubmitting = ref(false)
```
[the remark this answers is not in the document]
Comment (martinmaillard) [reply ui/src/components/JobForm.vue:59]: It makes the property reactive. It's part of the new [composition API](https://vuejs.org/api/composition-api-setup.html)

File: ui/src/components/RadioButton.vue [hunk 4]
```diff
@@ -12,12 +12,12 @@
...
 
 <script lang="ts">
-import { defineComponent } from '@vue/composition-api'
+import { defineComponent } from 'vue'
 
 export default defineComponent({
   name: 'RadioButton',
   props: {
-    value: {
+    modelValue: {
```
[the remark this answers is not in the document]
Comment (martinmaillard) [reply ui/src/components/RadioButton.vue:20]: It's the new convention for `v-model` in Vue 3

File: ui/tsconfig.json [hunk 5]
```diff
@@ -13,6 +13,7 @@
     "useDefineForClassFields": false,
     "sourceMap": true,
     "baseUrl": ".",
+    "downlevelIteration": true,
```
[the remark this answers is not in the document]
Comment (martinmaillard) [reply ui/tsconfig.json:16]: Very good question. The e2e tests wouldn't run without this because of `[...dataset]` I think. I spent a bit of time on it but not enough to really understand the issue.

File: ui/src/forms/editors/DictionaryTableEditor.vue (hunk 1 above)
Comment (martinmaillard) [reply ui/src/forms/editors/DictionaryTableEditor.vue:46]: In order to make the `templateResult` a computed property instead of a function call in the template itself. The function call was causing the template to re-render in a loop.

File: ui/src/forms/vue-custom-element.ts (hunk 2 above)
Comment (martinmaillard) [reply ui/src/forms/vue-custom-element.ts:4]: Because of:
1. The option to not create a shadow DOM (in order for the app styles to apply to the components)
2. Setting the app context on the components (otherwise, they don't have access to things that are configured on the app itself. Mostly Oruga components and config)

These 2 things basically replicate the behavior we had with `vue-custom-element`. I think [this comment](https://github.com/karol-f/vue-custom-element/issues/259#issuecomment-933008632) summarizes the situation.

File: ui/src/components/CubeProjectsItem.vue [hunk 6]
```diff
@@ -27,7 +27,9 @@
         </o-tooltip>
       </div>
     </div>
-    <o-collapse animation="slide" :open.sync="detailsShown">
+    <o-collapse animation="slide" :open="detailsShown">
+      <!-- A bug in Oruga makes the trigger slot mandatory -->
```
[the remark this answers is not in the document]
Comment (martinmaillard) [reply ui/src/components/CubeProjectsItem.vue:31]: I didn't report the issue, no

**Comment by martinmaillard** [comment]
@tpluscode your error should be fixed.

It would make sense to add a changelog entry for such a big change, wouldn't it? Event though it's not supposed to affect users...

**Review by tpluscode — APPROVED** [review]
No, I don't think I would add a chengelog entry.

**Review by tpluscode — COMMENTED** [review]
Actually, having checked out the branch I notice that IDE complains about `this.$store` and `this.$route`

This SO answer helps: https://stackoverflow.com/a/64414117/1103498

**Comment by tpluscode** [comment]
Ah, wait. We already have this in `vuex.d.ts` but it's not getting picked up.

**Comment by martinmaillard** [comment]
It doesn't complain about it for me, I already declared `$store` in `ui/types/vuex.d.ts` and vue-router already declares `$router`.

**Comment by tpluscode** [comment]
For me it works better if the declarations are moved to `shims-vue.d.ts`

**Comment by martinmaillard** [comment]
It's weird that it would matter, since these types files are explicitly declared in tsconfig.json. But I've also had issues with this in the past...

**Comment by martinmaillard** [comment]
Should I rename all the types files to `shims-something`?

**Comment by martinmaillard** [comment]
I'll merge this and we can address that in a separate PR if needed
